### PR TITLE
[SPARK-27937][CORE] Revert partial logic for auto namespace discovery

### PR DESCRIPTION
**Pull request by dhruve**
## What changes were proposed in this pull request?
This change reverts the logic which was introduced as a part of SPARK-24149 and a subsequent followup PR. 

With existing logic:
- Spark fails to launch with HDFS federation enabled while trying to get a path to a logical nameservice.  
- It gets tokens for unrelated namespaces if they are used in HDFS Federation
- Automatic namespace discovery is supported only if these are on the same cluster.

Rationale for change:
- For accessing data from related namespaces, viewfs should handle getting tokens for spark
- For accessing data from unrelated namespaces(user explicitly specifies them using existing configs) as these could be on the same or different cluster.

(Please fill in changes proposed in this fix)
Revert the changes.

## How was this patch tested?
Ran few manual tests and unit test.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
It seems to fail at UT, @dhruve . If this is a right fix, please make this pass the tests.

**Comment by dongjoon-hyun** [comment]
cc @vanzin

**Comment by dongjoon-hyun** [comment]
Also cc @mgaido91 since this wants to revert `SPARK-24149`.

**Comment by dhruve** [comment]
cc @attilapiros

**Comment by mgaido91** [comment]
I am not sure this is the right approach.

First, this change introduces a breaking change which can introduce regressions with federated namespaces: jobs running with the current code may break if they rely on this behavior after this PR.

Moreover:

> For accessing data from related namespaces, viewfs should handle getting tokens for spark

This is already the case, please see the comment in the code you reverted. If there is a case when this is not true, we should rather fix it than revert the full change IMHO.

> For accessing data from unrelated namespaces(user explicitly specifies them using existing configs) as these could be on the same or different cluster.

This is still the case, if you want to access a namespace which is on a different cluster, you still have to use that config.

Probably I am missing something, may you please provide a concrete use case where the current behavior is causing issues and which are the issues? Thanks.

**Comment by dhruve** [comment]
I have updated the jira with the use case where the current behavior of spark fails to launch an application.

- If you get a token from a specific NN, it is valid with or without HA. HDFS takes care of this. I don't think that Spark needs to do this explicitly. May be an hdfs expert can give a better perspective. cc @kihwal @daryn-sharp

- From the discussion in SPARK-24149, the reason why this change was introduced seems to address a very specific use case (while it is common to have tables and partitions across different namespaces, this one addresses a very specific use case: where partitions are stored across unrelated namespaces, using federation and luckily on the same cluster ) which I am not sure of is very common.

> the use case for this change, for instance, is when you have a partitioned table, when the partitions are on different namespaces and there is no viewFS configured

**Comment by mgaido91** [comment]
@dhruve what you are calling a very specific use case, is what in the hadoop documentation is called federation: https://hadoop.apache.org/docs/r2.8.3/hadoop-project-dist/hadoop-hdfs/Federation.html#Federation_Configuration.

Thanks for updating the JIRA. I am yet not very clear what the problem is in your case: I mean what was your configuration (`hdfs-site.xml`)? Anyway, why isn't setting `spark.kerberos.access.hadoopFileSystems` a solution for your use case? With that you can basically decide what to do and all the logic here is skipped.

**Comment by dhruve** [comment]
I think I have tried to explain the use case in the linked jira/s, but it seems it isn't clear. 

Let me try to explain the use case for which the change was implemented as described in the original jira:

Storing table/hive partitions across namespaces isn't uncommon. For storing partitions of a given table across namespaces, a good design would dictate that storing related data in related namespaces. This can be either in the same namespace or a large namespace that is broken into sub namespaces for better namespace management. In this case, good design would dictate to use viewfs. Just because we have the option of selecting a namespace while creating a hive partition, choosing an unrelated namespace is a poor decision. While a single team or organization might do this for whatever reason, I don't think this is the norm or hadoop/hive encourage storage policies where to store related data in unrelated namespaces, especially when the data belongs to the same table.

Because of Federation, the unrelated namespace happens to be stored on the same cluster as the other namespace. This, is okay.

So the problem we wanted to achieve comes down to:
Can we make spark figure it out for us, so the user doesn't need to know about the storage choices. 

Question is how do we do this. We cannot do this, because spark doesn't know anything about how the data is stored across namespaces(this is where using viewfs would make total sense). 

The use case which I refer to - Just because the data ended up being on the same cluster, we figure out the namespace information from the federation configs.
Some issues that I see with the approach:

1. It doesn't solve the issue. The user still has to know about the storage choices. If the data is stored on a different cluster, this solution doesn't work. Its just a hack to get around because of a poor design choice.
2. We now get tokens for namespaces which the user isn't going to read/write from. This was broken and there was another patch on top of the change to avoid doing exactly that and mention namespaces explicitly
 - (https://github.com/apache/spark/commit/c3f285c939ba046de5171ada9c4bbb1a2589635d)
3. If hadoop already figures out getting tokens for different namespaces uing viewfs, that is a better choice.
4. It fails to launch spark with existing HDFS deployments while trying to create a path from NameServiceIDs.

**Comment by mgaido91** [comment]
Answering to your points:

> If the data is stored on a different cluster, this solution doesn't work.

Well, that's quite expected since Spark runs on a specific hadoop cluster. For that case, you still have to do it manually, as before the patch you'd like to revert. So, no change on this.

> We now get tokens for namespaces which the user isn't going to read/write from.

As you mentioned, there is also a configuration for that. So if you want, you can limit your app to get only the tokens you need.

> If hadoop already figures out getting tokens for different namespaces using viewfs, that is a better choice.

Yes, but this is not related with Spark code and this patch, it is rather related to how you configure and use your cluster. You're definitely allowed to use viewfs and no extra token is taken in that case...

> It fails to launch spark with existing HDFS deployments while trying to create a path from NameServiceIDs.

Then this is a bug, if you have a fix for this, that's great.

**Comment by dhruve** [comment]
This is a bug, and we should fix it. Taking a closer look, we don't need the change at all. Just specify an additional namespace - you don't need one more config, it also doesn't break anything.

**Comment by daryn-sharp** [comment]
The client should specify all the required filesystems directly or indirectly via viewfs.  It’s a very user specific use case to assume that not using viewfs means the config should be scanned for all federated/HA namespaces.

A client should never distinguish between filesystem types when acquiring tokens.  Viewfs just happens to be a handy way to implicitly collect tokens for all mounted filesystems rather than explicitly specifying them but it should not be the basis for any conditionalized behavior.

Federation is a server-side only configuration for enabling multi-namespace DNs.  A client cannot and should not distinguish between a federated and non-federated namespace.  Technically a federation config doesn’t even belong in a client config and as such should never be used by a client.

In short, the user’s use case is better solved specifying the filesystems.  If the federation configs are leaked into the client configs, it should be easy to replace them with a conf key of required filesystems.

**Comment by dhruve** [comment]
Thanks @daryn-sharp for the inputs. @mgaido91 - hope that explanation gives more perspective.

**Comment by mgaido91** [comment]
I see your point and I respect it. Despite this, reverting that change has breaking effects, so it should be done only if there are strong reasons for that, ie. use cases which are not working in the current way, which isn't the case IMHO, as there is also a config to specify the namespaces strictly required.

**Comment by HyukjinKwon** [comment]
Guys, sorry but can anyone clarify if:

1. Did the auto namespace discovery cause any regression?
2. If so, is it a corner case, or is it something worthy reverting the HDFS federation out of the box?
 
Sounds like we rather better fix rather than revert.

**Comment by dhruve** [comment]
It did cause a regression in our [Yahoo/VerizonMedia] clusters. And the way spark currently implements them is totally unnecessary. You can read the context from the jira and also the comments above. The change is a hack around a very user specific case and ends up relying on reading hdfs server configs, which is not what any client should be doing.

**Comment by dhruve** [comment]
It can also lead to confusion for the user, because if a table is stored across multiple namespaces, the user will see two behaviors:

1. If federation is enabled:
      a. If data is stored on same cluster => User can query the data
      b. If data is stored on diff cluster => User has to specify the namespace explicitly
2. If federation is disabled:
     User has to specify the namespace explicitly

Before the change:
User has to specify the namespace explicitly all the times. Which I think is more consistent to reason about.

**Comment by vanzin** [comment]
> Technically a federation config doesn’t even belong in a client config and as such should never be used by a client.

Well, technically you have similar configuration on the client so that the viewfs mount points can reference the appropriate namespaces... but that's nitpicking.

But in general I agree that it feels weird for the client to be looking at random HDFS configurations (in fact I made a similar comment on the original review and requested to special-case viewfs if the config scraping code was going to be added).

Given that I'm ok with removing the config scraping code. Viewfs users (which in my view should be all users accessing a federated HDFS) shouldn't see any difference. Those using the Spark config should not see any difference either. Only those relying on a weird HDFS config that had multiple configured namespaces but not viewfs should see a difference... but well, it's the price to pay if this causes problems for people.

(I haven't looked at the code yet.)

File: docs/running-on-yarn.md [hunk 1]
```diff
@@ -553,12 +553,9 @@ For example, suppose you would like to point log url link to Job History Server
...
 Standard Kerberos support in Spark is covered in the [Security](security.html#kerberos) page.
 
-In YARN mode, when accessing Hadoop filesystems, Spark will automatically obtain delegation tokens
-for:
-
-- the filesystem hosting the staging directory of the Spark application (which is the default
-  filesystem if `spark.yarn.stagingDir` is not set);
-- if Hadoop federation is enabled, all the federated filesystems in the configuration.
+In YARN mode, when accessing Hadoop file systems, aside from the service hosting the user's home
+directory, Spark will also automatically obtain delegation tokens for the service hosting the
```
Comment (vanzin) [inline docs/running-on-yarn.md]: Instead of "service hosting the user's home directory", the code is actually using the default file system from the configuration.

**Comment by vanzin** [comment]
retest this please

File: docs/running-on-yarn.md (hunk 1 above)
Comment (dhruve) [reply docs/running-on-yarn.md]: I will update the documentation. I missed it. This was just a revert to the old configuration.

Comment (vanzin) [reply docs/running-on-yarn.md]: Still waiting...

Comment (dhruve) [reply docs/running-on-yarn.md]: I have updated the docs.

**Comment by vanzin** [comment]
Merging to master.

**Comment by dhruve** [comment]
Thanks @vanzin
